napari/utils/lock.py

```diff
@@ -0,0 +1,215 @@
+"""
+A generic lock class to be used as a locking mechanism.
+"""
+from enum import Enum
+from typing import Any, Dict, List, Optional
+
+from pydantic import BaseModel, validator
+
+from napari.utils.translations import trans
+
+
+class LockMode(Enum):
+    """Lock modes"""
+
+    EXACT = 0
+    IN_LIST = 1
+    IN_RANGE = 2
+    LARGER_THAN = 3
+    SMALLER_THAN = 4
+
+
+class ValueNotCompatibleWithLockMode(Exception):
+    """Custom error that is raised when the value is not compatible with lock mode"""
+
+    def __init__(self, value, message: str) -> None:
+        self.value = value
+        self.message = f"value: {value}. " + message
+        super().__init__(message)
+
+
+class AttributeNotFound(Exception):
+    """Custom error that is raised when the attribute is not found in the locker"""
+
+    def __init__(self, message: str) -> None:
+        self.message = message
+        super().__init__(message)
+
+
+class Lock(BaseModel):
+    """
+    A lock consists of components:
+    1. A boolean to house the status of the lock (open, locked).
+    2. A value of the lock, for example, an attribute maximum can be locked to certain number.
+    3. A lock mode to desginate the type of lock (soft (user), or hard (code and user))
+    4. comments
+    """
+
+    value_lock_mode: LockMode = LockMode.EXACT
+    value: Any = None
+    locked: bool = True
+    hard_lock: bool = True
+    comments: Optional[str] = None
+
+    @validator("value_lock_mode")
+    def value_lock_mode_lock_mode_valid(cls, value):
+        """Validate value lock mode"""
+        return value
+
+    @validator("value")
+    def value_lock_mode_valid(cls, value, values, **kwargs):
+        """Validate value"""
+        value_lock_mode = values["value_lock_mode"]
+
+        def is_numeric(x):
+            """Quick checking if it is numeric type"""
+            return type(x) in (int, float)
```

Comment:
You may want to make that a utils function outside of the method to no redefine it each time.

Reply:
removed the function as it turned out it is only being used once.